Unique factorization domains

tirix opened the pull request:
Adds a few theorems about unique factorization domains (UFD), especially, `~1arithufd`, the existence of a factorization into primes. Since the definition in `set.mm` is based on Kaplansky's characterization, this is not the definition but a property.

savask commented on set.mm:
```diff
@@ -512846,6 +512945,25 @@ commutative monoid (=vectors) together with a semiring (=scalars) and a
+    ringdi22.6 $e |- ( ph -> Y e. B ) $.
+    ringdi22.7 $e |- ( ph -> Z e. B ) $.
+    ringdi22.8 $e |- ( ph -> T e. B ) $.
+    $( Develop the product of two sums in a ring.  (Contributed by Thierry
```
I think "Expand the product" is a better term.

tirix replied:
Fixed in https://github.com/metamath/set.mm/pull/4872/commits/cc2150ead40d9ee3851e56d8f37942b85b152b14

savask commented on set.mm:
```diff
@@ -520296,6 +520699,33 @@ such that every nonzero prime ideal contains a prime element (this is a
+    ufdprmidl.2 $e |- ( ph -> R e. UFD ) $.
+    ufdprmidl.3 $e |- ( ph -> J e. I ) $.
+    ufdprmidl.4 $e |- ( ph -> J =/= { .0. } ) $.
+    $( In a unique factorization domains ` R ` , a nonzero prime ideal ` J `
```
"domains" -> "domain"

savask commented on set.mm:
```diff
@@ -520296,6 +520699,33 @@ such that every nonzero prime ideal contains a prime element (this is a
+    $d R i $.
+    ufdidom.1 $e |- ( ph -> R e. NzRing ) $.
+    ufdidom.2 $e |- ( ph -> R e. UFD ) $.
+    $( A nonzero UFD is an integral domain.  (Contributed by Thierry Arnoux,
```
Maybe it's better to expand "UFD" into "unique factorization domain" here, since you tend to use the non-abbreviated term in other theorems.

savask commented on set.mm:
```diff
@@ -520332,6 +520762,300 @@ such that every nonzero prime ideal contains a prime element (this is a
+          1arithufdlem3.y $e |- ( ph -> Y e. B ) $.
+          1arithufdlem3.1 $e |- ( ph -> ( Y .x. X ) e. S ) $.
+          $( Lemma for ~ 1arithufd .  If a product ` ( Y .x. X ) ` can be
+             written as a product of primes, with ` X ` non-unit, nonzero, so
```
This is a very minor nitpick which can be safely ignored, but it's better to choose whether we hyphenate non- or not. I.e. write "nonunit, nonzero" or "non-unit, non-zero".

tirix replied:
In https://github.com/metamath/set.mm/pull/4872/commits/cc2150ead40d9ee3851e56d8f37942b85b152b14, I chose to use "non-unit, non-zero", however I'm open to any suggestion.

savask reviewed (approved):
All very minor nitpicks, and nice work as always!

I must admit I have already lost a sense of direction a little bit, is this work on UFDs related to the doubling the cube work?

tirix commented:
Note that in the [last commit](https://github.com/metamath/set.mm/pull/4872/commits/cc2150ead40d9ee3851e56d8f37942b85b152b14), following very pertinent remarks from @langgerard, I also deleted `~ drngnidlr` which was in fact redundant with `~ drngidl` (while weaker!), and removed commutativity hypotheses from `~ drngmxidlr` and `~ krullndrng`. I don't know how to best attribute this contribution to Gérard, since it does not make sense to add a "Proof modification" tag.

tirix commented:
> I must admit I have already lost a sense of direction a little bit, is this work on UFDs related to the doubling the cube work?

Maybe I also lost direction myself!

Next step is to prove that the polynomial ` ( ( X ^ 3 ) - 2 ) ` is indeed the minimal polynomial for ` ( 2 ^c ( 1 / 3 ) ) ` _when considering rationals_.
By `~ irredminply`, it suffices to prove that it's irreducible.

The standard proof of irreducibility uses reduction of the coefficients modulo a prime, either using _Eisenstein's Criterion_ or directly. Then in order to move from irreducibility over the integers to irreducibility over the rationals, the standard proof uses _Gauss's lemma_.

Both Eisenstein's Criterion and Gauss's lemma have formulations with UFD: this is why I've been proving theorems around irreducibles, prime ideals, prime elements and UFD recently.

savask commented:
> The standard proof of irreducibility uses reduction of the coefficients modulo a prime, either using Eisenstein's Criterion or directly.

You could also prove that a cubic polynomial (over any field) is irreducible if and only if it has no roots. Then the claim for x^3 - 2 would follow from [rtprmirr](https://us.metamath.org/mpeuni/rtprmirr.html). The UFD work is still very important regardless of the proof strategy you choose, though.

tirix commented:
> You could also prove that a cubic polynomial (over any field) is irreducible if and only if it has no roots. Then the claim for x^3 - 2 would follow from [rtprmirr](https://us.metamath.org/mpeuni/rtprmirr.html). The UFD work is still very important regardless of the proof strategy you choose, though.

Yes, that would work! Thanks!

wlammen commented:
> I don't know how to best attribute this contribution to Gérard,

What about using the comment section, as in [ax13lem1](https://us.metamath.org/mpeuni/ax13lem1.html)?

Thanks to Gerard Lang, whose ideas allowed several improvements.
